### PR TITLE
Add verify_backward to enable testing bacward ops

### LINKED ISSUE #1356: Add verification for backward pass into verify function
**Proposed solution**

1. calculate forwards (this could be reused from verify forward)
2. generate random gradient for the input in backward
3. calculate gradients for all parametars (run framework and compiled backward)
4. verify each gradient for: dtype, shape, value

### PR DESCRIPTION
**Pull request by ndrakulicTT**
### Ticket
Fixes #1356

### Problem description
There was no way to in one place verify gradients from backward pass

### What's changed
Added verify_backward:
- check input parameters
- runs backward on compile model
- saves gradients from compiled model
- runs bacward on framework model
- saves gradients from framework model
- Checks the gradients

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: forge/forge/verify/verify.py [hunk 1]
```diff
@@ -12,11 +12,18 @@
...
+from forge.tensor import to_pt_tensors, TensorFromTrace
 
 from ..tensor import Tensor, TensorShape, pad_pytorch_tensor_to_forge, narrow_forge_tensor_to_pytorch
-from .config import DepricatedVerifyConfig, VerifyConfig, VerifyTensorMetadata, should_waive_gradient
+from .config import (
+    DepricatedVerifyConfig,
+    VerifyConfig,
+    VerifyTensorMetadata,
+    should_waive_gradient,
+    AutomaticValueChecker,
```
Comment (vkovinicTT) [inline forge/forge/verify/verify.py:25]: Please just remove `import AutomaticValueChecker ` if we don't use it here?

File: forge/test/mlir/llama/tests/test_specific_ops_llama32.py [hunk 2]
```diff
@@ -153,14 +162,36 @@ def forward(self, x, y):
...
-    compiled_model = forge.compile(framework_model, sample_inputs=inputs)
-
-    verify(inputs, framework_model, compiled_model, VerifyConfig(value_checker=AutomaticValueChecker(pcc=0.95)))
+    framework_model.eval() if not train else framework_model.train()
+
+    # NOTE: We probably need two framework models with the same state_dict to compare the outputs
+    #       But for now it works without that for some reason?
+    # model_for_compile = Matmul()
+    # model_for_compile.eval() if not training else model_for_compile.train()
+    # model_for_compile.load_state_dict(framework_model.state_dict())
```
Comment (vkovinicTT) [inline forge/test/mlir/llama/tests/test_specific_ops_llama32.py:176]: Discussed offline, we probably don't need 2 framework models.

File: forge/forge/verify/verify.py [hunk 3]
```diff
@@ -302,6 +302,103 @@ def check_dtypes(fw_dtype: torch.dtype, co_dtype: torch.dtype):
...
+    framework_model.zero_grad()
+
+    # 1st step: run backward pass for the networks and get gradients
+    compiled_model.gradient_inputs = [output_grad]
+    co_gradient_outputs = compiled_model.backward()
+    co_gradients: Dict[str, torch.Tensor] = {}
+    for name, grad in zip(compiled_model.bwd_compiled_graph_state.ordered_output_names, co_gradient_outputs):
+        # NOTE: Need to clone the gradients of parametars as they are modified in the backward pass of the framework model
+        #       but no need to clone the gradients of the inputs as they are not modified in the backward pass of the framework model
+        co_gradients[name] = grad.clone() if name.startswith("grad_acc_") else grad
```
Comment (vkovinicTT) [inline forge/forge/verify/verify.py:367]: Will this be issue if we try to verify larger models?

File: forge/forge/verify/verify.py [hunk 4]
```diff
@@ -302,6 +302,103 @@ def check_dtypes(fw_dtype: torch.dtype, co_dtype: torch.dtype):
...
+            name = name.replace("grad_acc_", "")
+            name = name.replace("_grad_accumulator", "")
+            fw_grad = framework_model.get_parameter(name).grad
+        elif name.startswith("output_grad_"):
+            name = name.replace("output_grad_", "")
+            fw_grad = inputs[compiled_model.fwd_compiled_graph_state.ordered_input_names.index(name)].grad
+        else:
+            raise ValueError(f"Unknown gradient name in compiled model: {name}")
+
+        assert co_grad.data_ptr() != fw_grad.data_ptr(), "Gradients are the same object in memory"
```
Comment (vkovinicTT) [inline forge/forge/verify/verify.py:387]: is this assertion because of [line 367](https://github.com/tenstorrent/tt-forge-fe/pull/1383/files#diff-e20a777f7a22e33ebefaedc11a005cbb40c7ca37adbe814a5d84baeae45e775bR367)?

File: forge/forge/verify/verify.py (hunk 3 above)
Comment (ndrakulicTT) [reply forge/forge/verify/verify.py:367]: Gradients are of the size of the model parameters so it will effectively be one more model loaded not sure if that is too much

Comment (vkovinicTT) [reply forge/forge/verify/verify.py:367]: I think it's fine because if a mismatch occurs in the model, we can dissect it and verify the backward pass for individual submodules.

**Review by vkovinicTT — APPROVED** [review]

File: forge/test/mlir/llama/tests/test_specific_ops_llama32.py [hunk 5]
```diff
@@ -128,21 +128,23 @@ def forward(self, x):
...
+        (((32, 11, 11), (32, 11, 64)), False),
+        (((11, 2048), (2048, 8192)), False),
+        (((1, 11, 8192), (8192, 2048)), False),
+        (((1, 11, 2048), (2048, 128256)), False),
+        (((32, 11, 64), (32, 64, 11)), True),
     ],
 )
 @pytest.mark.push
-def test_matmul(shapes):
+def test_matmul(shapes, train):
```
Comment (pilkicTT) [inline forge/test/mlir/llama/tests/test_specific_ops_llama32.py:147]: we should probably add a pytest fixture for `mode = Inference | Training`
but let's leave that for another day...

**Review by pilkicTT — APPROVED** [review]

**Review by vladimirjovanovicTT — APPROVED** [review]

File: forge/forge/verify/verify.py [hunk 6]
```diff
@@ -281,6 +281,103 @@ def check_dtypes(fw_dtype: torch.dtype, co_dtype: torch.dtype):
...
+
+    # Run backward on framework model
+    framework_model.zero_grad()
+    framework_output.backward(gradient=output_grad)
+
+    # 2nd step: verify gradients
+    for name in co_gradients:
+        co_grad = co_gradients[name]
+
+        if name.startswith("grad_acc_"):
```
[the remark this answers is not in the document]
Comment (ndrakulicTT) [reply forge/forge/verify/verify.py:356]: Here is the issue that should address this: #1519

[the remark this answers is not in the document]
Comment (vladimirjovanovicTT) [reply forge/forge/verify/verify.py:356]: Great!
